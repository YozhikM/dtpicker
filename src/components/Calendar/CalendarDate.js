/* @flow */
/* eslint-disable jsx-a11y/no-static-element-interactions */

import React from 'react';
import { setMonth, setYear, addYears, addMonths, getYear, format } from 'date-fns';
import TableSelect from '../TableSelect/TableSelect';
import CalendarMonthGrid from './CalendarMonthGrid';
import TimePicker from './TimePicker';
import SvgIcon from '../SvgIcon';
import s from './Calendar.scss';

type Show = 'calendar' | 'mm' | 'yy' | 'yy10';

export type Props = {|
  value: Date,
  onChange?: (value: Date) => void, // works on change day, month, year, year10, time (it only work for displaying value)
  highlight?: Array<Date> | Date,
  visibleTime?: Date,
  onSetDate?: (date: Date) => void, // user click by day in calendar
  onSetTime?: (time: Date) => void, // user click on time in calendar
  leftArrow?: boolean, // show left arrow in header (true by default)
  rightArrow?: boolean, // show right arrow in header (true by default)
  time?: boolean, // display time picker (hidden by default)
  show?: Show // display needed view type (calendar by default)
|};

type State = {|
  show: 'calendar' | 'mm' | 'yy' | 'yy10',
  value: Date,
  monthsOptions: Array<{ value: number, name: string }>
|};

class CalendarDate extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      show: this.props.show || 'calendar',
      value: this.props.value,
      monthsOptions: this.getMonthOptions()
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value
      });
    }
  }

  onSetDate = (date: Date) => {
    const { onSetDate } = this.props;
    if (onSetDate) onSetDate(date);
  };

  onSetTime = (time: Date) => {
    const { onSetTime } = this.props;
    if (onSetTime) onSetTime(time);
  };

  onChange = (value: Date) => {
    this.setState({ value }, () => {
      const { onChange } = this.props;
      if (onChange) onChange(value);
    });
  };

  onChangeMonth = (n: number) => {
    const { value } = this.state;
    const newDate = setMonth(value, n);
    this.setState({ show: 'calendar' });
    this.onChange(newDate);
  };

  onChangeYear = (n: number) => {
    const { value } = this.state;
    const newDate = setYear(value, n);
    this.setState({ show: 'mm' });
    this.onChange(newDate);
  };

  onChangeDecade = (n: number) => {
    const { value } = this.state;
    const newDate = setYear(value, n);
    this.setState({ show: 'yy' });
    this.onChange(newDate);
  };

  showCalendar = () => {
    this.setState({ show: 'calendar' });
  };

  showMonthTable = () => {
    this.setState({ show: 'mm' });
  };

  showYearTable = () => {
    this.setState({ show: 'yy' });
  };

  showDecadeTable = () => {
    this.setState({ show: 'yy10' });
  };

  incrementMonth = () => {
    const value = addMonths(this.state.value, 1);
    this.onChange(value);
  };

  decrementMonth = () => {
    const value = addMonths(this.state.value, -1);
    this.onChange(value);
  };

  increment10Years = () => {
    const { value } = this.state;
    const newDate = addYears(value, 10);
    this.onChange(newDate);
  };

  decrement10Years = () => {
    const { value } = this.state;
    const newDate = addYears(value, -10);
    this.onChange(newDate);
  };

  incrementYears = () => {
    const { value } = this.state;
    const newDate = addYears(value, 1);
    this.onChange(newDate);
  };

  decrementYears = () => {
    const { value } = this.state;
    const newDate = addYears(value, -1);
    this.onChange(newDate);
  };

  showDecade = () => {
    const { value } = this.state;
    const tmp = getYear(value);
    const startYear = tmp - tmp % 10;
    const decade = `${startYear} - ${startYear + 9}`;
    return decade;
  };

  getMonthOptions() {
    return [...Array(12)].map((v, i) => ({
      value: i,
      name: format(new Date(2000, i, 1), 'MMM')
    }));
  }

  getYearOptions(date: Date) {
    const tmp = date.getFullYear();
    const startYear = tmp - tmp % 10;
    return [...Array(10)].map((v, i) => ({
      value: startYear + i,
      name: `${startYear + i}`
    }));
  }

  getYearDecadeOptions(date: Date, cnt: number = 5) {
    const tmp = date.getFullYear();
    const middleYear = tmp - tmp % 10;
    const startYear = middleYear - Math.ceil(cnt / 2) * 10;
    return [...Array(cnt)].map((v, i) => {
      const year = startYear + i * 10;
      return {
        value: year,
        name: `${year} ${year + 9}`
      };
    });
  }

  render() {
    const { show, value, monthsOptions } = this.state;
    const { time, highlight, visibleTime } = this.props;
    const leftArrowStyle = {
      visibility: this.props.leftArrow === false ? 'hidden' : 'visible'
    };
    const rightArrowStyle = {
      visibility: this.props.rightArrow === false ? 'hidden' : 'visible'
    };

    if (show === 'yy10') {
      const curYear = getYear(value);
      return (
        <div className={s.calendar_container}>
          <div className={s.table}>
            <span>
              &nbsp; <br /> &nbsp;
            </span>
            <TableSelect
              options={this.getYearDecadeOptions(value, 8)}
              cols={4}
              value={curYear - curYear % 10}
              onChange={this.onChangeDecade}
            />
            <button onClick={this.showYearTable}>Назад</button>
          </div>
        </div>
      );
    }

    if (show === 'yy') {
      return (
        <div className={s.calendar_container}>
          <div className={s.table}>
            <div className={s.c_button_container}>
              <button onClick={this.decrement10Years}>
                <SvgIcon file="arrow-left" />
              </button>
              <span onClick={this.showDecadeTable}>{this.showDecade()}</span>
              <button onClick={this.increment10Years}>
                <SvgIcon file="arrow-right" />
              </button>
            </div>
            <TableSelect
              options={this.getYearOptions(value)}
              cols={4}
              value={getYear(value)}
              onChange={this.onChangeYear}
            />
            <button onClick={this.showMonthTable}>Назад</button>
          </div>
        </div>
      );
    }

    if (show === 'mm') {
      return (
        <div className={s.calendar_container}>
          <div className={s.table}>
            <div className={s.c_button_container}>
              <button onClick={this.decrementYears}>
                <SvgIcon file="arrow-left" />
              </button>
              <span onClick={this.showYearTable}>{getYear(value)}</span>
              <button onClick={this.incrementYears}>
                <SvgIcon file="arrow-right" />
              </button>
            </div>
            <TableSelect
              options={monthsOptions}
              cols={4}
              value={value.getMonth()}
              onChange={this.onChangeMonth}
            />
            <button onClick={this.showCalendar}>Назад</button>
          </div>
        </div>
      );
    }

    return (
      <div className={s.calendar_container}>
        <button style={leftArrowStyle} className={s.left_arrow} onClick={this.decrementMonth}>
          <SvgIcon file="arrow-left" />
        </button>
        <button style={rightArrowStyle} className={s.right_arrow} onClick={this.incrementMonth}>
          <SvgIcon file="arrow-right" />
        </button>
        <CalendarMonthGrid
          highlight={highlight}
          value={value}
          onClickMonth={this.showMonthTable}
          onSetDate={this.onSetDate}
        />
        {time && <TimePicker date={visibleTime} onSetTime={this.onSetTime} />}
      </div>
    );
  }
}

export default CalendarDate;
