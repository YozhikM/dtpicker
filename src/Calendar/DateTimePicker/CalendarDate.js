/* @flow */

import React from 'react';
import { setMonth, setYear, addYears, addMonths, getYear, format } from 'date-fns';
import SVGInline from 'react-svg-inline';
import Button from '../../Button/Button';
import SvgIcon from '../../SvgIcon/SvgIcon';
import TableSelect from '../TableSelect/TableSelect';
import CalendarMonthGrid from './CalendarMonthGrid';
import TimePicker from './TimePicker';
import './Calendar.scss';
import arrowLeft from './arrow-left.svg';
import arrowRight from './arrow-right.svg';

type Show = 'calendar' | 'mm' | 'yy' | 'yy10';

export type Props = {|
  value: Date,
  onChange?: (value: Date) => void, // works on change day, month, year, year10, time (it only work for displaying value)
  highlight?: Array<Date> | Date,
  visibleTime?: Date,
  onSetDate?: (date: Date) => void, // user click by day in calendar
  onSetTime?: (time: Date) => void, // user click on time in calendar
  leftArrow?: boolean, // show left arrow in header (true by default)
  rightArrow?: boolean, // show right arrow in header (true by default)
  time?: boolean, // display time picker (hidden by default)
  show?: Show, // display needed view type (calendar by default)
  minDate?: Date,
  maxDate?: Date,
|};

type State = {|
  show: 'calendar' | 'mm' | 'yy' | 'yy10',
  value: Date,
  monthsOptions: Array<{ value: number, name: string }>,
|};

class CalendarDate extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      show: props.show || 'calendar',
      value: props.value,
      monthsOptions: this.getMonthOptions(),
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.value !== nextProps.value) {
      this.setState({
        value: nextProps.value,
      });
    }
  }

  onSetDate = (date: Date) => {
    const { onSetDate } = this.props;
    if (onSetDate) onSetDate(date);
  };

  onSetTime = (time: Date) => {
    const { onSetTime } = this.props;
    if (onSetTime) onSetTime(time);
  };

  onChange = (value: Date) => {
    this.setState({ value: new Date() }, () => {
      const { onChange } = this.props;
      if (onChange) onChange(value);
    });
  };

  onChangeMonth = (n: number) => {
    const { minDate, maxDate } = this.props;
    const { value } = this.state;
    const newDate = setMonth(value, n);

    if ((maxDate && newDate > maxDate) || (minDate && newDate < minDate)) return;

    this.setState({ show: 'calendar' });
    this.onChange(newDate);
  };

  onChangeYear = (n: number) => {
    const { minDate, maxDate } = this.props;
    const { value } = this.state;
    const newDate = setYear(value, n);

    if ((maxDate && newDate > maxDate) || (minDate && newDate < minDate)) return;

    this.setState({ show: 'mm' });
    this.onChange(newDate);
  };

  onChangeDecade = (n: number) => {
    const { minDate, maxDate } = this.props;
    const { value } = this.state;
    const newDate = setYear(value, n);

    if ((maxDate && newDate > maxDate) || (minDate && newDate < minDate)) return;

    this.setState({ show: 'yy' });
    this.onChange(newDate);
  };

  showCalendar = () => {
    this.setState({ show: 'calendar' });
  };

  showMonthTable = () => {
    this.setState({ show: 'mm' });
  };

  showYearTable = () => {
    this.setState({ show: 'yy' });
  };

  showDecadeTable = () => {
    this.setState({ show: 'yy10' });
  };

  incrementMonth = () => {
    const { maxDate } = this.props;
    const value = addMonths(this.state.value, 1);
    if (maxDate && value > maxDate) return;

    this.onChange(value);
  };

  decrementMonth = () => {
    const { minDate } = this.props;
    const value = addMonths(this.state.value, -1);
    if (minDate && value < minDate) return;

    this.onChange(value);
  };

  increment10Years = () => {
    const { maxDate } = this.props;
    const { value } = this.state;
    const newDate = addYears(value, 10);
    if (maxDate && newDate > maxDate) return;

    this.onChange(newDate);
  };

  decrement10Years = () => {
    const { minDate } = this.props;
    const { value } = this.state;
    const newDate = addYears(value, -10);
    if (minDate && newDate < minDate) return;

    this.onChange(newDate);
  };

  incrementYears = () => {
    const { maxDate } = this.props;
    const { value } = this.state;
    const newDate = addYears(value, 1);
    if (maxDate && newDate > maxDate) return;

    this.onChange(newDate);
  };

  decrementYears = () => {
    const { minDate } = this.props;
    const { value } = this.state;
    const newDate = addYears(value, -1);
    if (minDate && newDate < minDate) return;

    this.onChange(newDate);
  };

  showDecade() {
    const { value } = this.state;
    const tmp = getYear(value);
    const startYear = tmp - tmp % 10;
    const decade = `${startYear} - ${startYear + 9}`;

    return decade;
  }

  getMonthOptions() {
    return [...Array(12)].map((v, i) => ({
      value: i,
      name: format(new Date(2000, i, 1), 'MMM'),
    }));
  }

  getYearOptions(date: Date) {
    const tmp = date.getFullYear();
    const startYear = tmp - tmp % 10;

    return [...Array(10)].map((v, i) => ({
      value: startYear + i,
      name: `${startYear + i}`,
    }));
  }

  getYearDecadeOptions(date: Date, cnt: number = 5) {
    const tmp = date.getFullYear();
    const middleYear = tmp - tmp % 10;
    const startYear = middleYear - Math.ceil(cnt / 2) * 10;

    return [...Array(cnt)].map((v, i) => {
      const year = startYear + i * 10;
      return {
        value: year,
        name: `${year} ${year + 9}`,
      };
    });
  }

  getBtnStyle() {
    const { leftArrow, rightArrow } = this.props;
    if (leftArrow && !rightArrow) {
      return { justifyContent: 'flex-start' };
    } else if (!leftArrow && rightArrow) {
      return { justifyContent: 'flex-end' };
    }
    return { justifyContent: 'space-between' };
  }

  render() {
    const { show, value, monthsOptions } = this.state;
    const { time, highlight, visibleTime, leftArrow, rightArrow, minDate, maxDate } = this.props;

    if (show === 'yy10') {
      const curYear = getYear(value);
      return (
        <div className="calendar_container">
          <div className="table">
            <span className="span">
              &nbsp; <br /> &nbsp;
            </span>
            <TableSelect
              options={this.getYearDecadeOptions(value, 8)}
              cols={4}
              value={curYear - curYear % 10}
              onChange={this.onChangeDecade}
            />
            <Button onClick={this.showYearTable} xs>
              Назад
            </Button>
          </div>
        </div>
      );
    }

    if (show === 'yy') {
      return (
        <div className="calendar_container">
          <div className="table">
            <div className="c_button_container">
              <button onClick={this.decrement10Years}>
                <SVGInline width="24" svg={arrowLeft} />
              </button>
              <span className="span" onClick={this.showDecadeTable}>
                {this.showDecade()}
              </span>
              <button onClick={this.increment10Years}>
                <SVGInline width="24" svg={arrowRight} />
              </button>
            </div>
            <TableSelect
              options={this.getYearOptions(value)}
              cols={4}
              value={getYear(value)}
              onChange={this.onChangeYear}
            />
            <Button onClick={this.showMonthTable} xs>
              Назад
            </Button>
          </div>
        </div>
      );
    }

    if (show === 'mm') {
      return (
        <div className="calendar_container">
          <div className="table">
            <div className="c_button_container">
              <button onClick={this.decrementYears}>
                <SVGInline width="24" svg={arrowLeft} />
              </button>
              <span className="span" onClick={this.showYearTable}>
                {getYear(value)}
              </span>
              <button onClick={this.incrementYears}>
                <SVGInline width="24" svg={arrowRight} />
              </button>
            </div>
            <TableSelect
              options={monthsOptions}
              cols={4}
              value={value.getMonth()}
              onChange={this.onChangeMonth}
            />
            <Button onClick={this.showCalendar} xs>
              Назад
            </Button>
          </div>
        </div>
      );
    }

    const isHiddenLeft = minDate && addMonths(value, -1) < minDate;
    const isHiddenRight = maxDate && addMonths(value, 1) > maxDate;

    return (
      <div className="calendar_container">
        <div className="btns" style={this.getBtnStyle()}>
          {leftArrow && (
            <button
              className="left_arrow"
              onClick={this.decrementMonth}
              style={isHiddenLeft ? { visibility: 'hidden' } : { cursor: 'pointer' }}
            >
              <SVGInline width="24" svg={arrowLeft} />
            </button>
          )}
          {rightArrow && (
            <button
              className="right_arrow"
              onClick={this.incrementMonth}
              style={isHiddenRight ? { visibility: 'hidden' } : { cursor: 'pointer' }}
            >
              <SVGInline width="24" svg={arrowRight} />
            </button>
          )}
        </div>
        <CalendarMonthGrid
          highlight={highlight}
          value={value}
          onClickMonth={this.showMonthTable}
          onSetDate={this.onSetDate}
          minDate={minDate}
          maxDate={maxDate}
        />
        {time && <TimePicker date={visibleTime} onSetTime={this.onSetTime} />}
      </div>
    );
  }
}

export default CalendarDate;
