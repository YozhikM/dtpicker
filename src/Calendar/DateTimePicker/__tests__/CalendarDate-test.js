/* @flow */

import React from 'react';
import { setMonth, setYear, addMonths, addYears } from 'date-fns';
import CalendarDate from '../CalendarDate';

describe('CalendarDate', () => {
  describe('render', () => {
    it('basic', () => {
      const value = new Date(2010, 10, 10);
      expect(
        shallow(<CalendarDate value={value} highlight={value} show="calendar" />)
      ).toMatchSnapshot();
    });

    it('if month table', () => {
      const value = new Date(2014, 11, 1);
      expect(shallow(<CalendarDate value={value} highlight={value} show="mm" />)).toMatchSnapshot();
    });

    it('if year table', () => {
      const value = new Date(2022, 0, 21);
      expect(shallow(<CalendarDate value={value} highlight={value} show="yy" />)).toMatchSnapshot();
    });

    it('if decade table', () => {
      const value = new Date(2040, 5, 10);
      expect(
        shallow(<CalendarDate value={value} highlight={value} show="yy10" />)
      ).toMatchSnapshot();
    });

    it('if time', () => {
      const value = new Date(2005, 6, 7);
      expect(shallow(<CalendarDate value={value} highlight={value} time />)).toMatchSnapshot();
    });

    it('if disabled leftArrow', () => {
      const value = new Date(2007, 6, 5);
      expect(
        shallow(<CalendarDate value={value} highlight={value} leftArrow={false} />)
      ).toMatchSnapshot();
    });

    it('if disabled rightArrow', () => {
      const value = new Date(1984, 4, 4);
      expect(
        shallow(<CalendarDate value={value} highlight={value} rightArrow={false} />)
      ).toMatchSnapshot();
    });
  });

  it('componentWillReceiveProps()', () => {
    const wrapper = shallow(<CalendarDate value={new Date()} highlight={new Date()} />);
    const nextProps = {
      value: new Date(1994, 7, 1),
      highlight: new Date(),
    };
    wrapper.instance().componentWillReceiveProps(nextProps);

    expect(wrapper.state('value')).toMatchSnapshot();
  });

  describe('user interactions', () => {
    it('onSetDate()', () => {
      const onSetDate = jest.fn();
      const wrapper = shallow(
        <CalendarDate highlight={new Date()} value={new Date()} onSetDate={onSetDate} />
      );
      const newDate = new Date(2000, 11, 11, 0, 0, 0);
      wrapper
        .find('CalendarMonthGrid')
        .props()
        .onSetDate(newDate);

      expect(onSetDate).toBeCalledWith(newDate);
    });

    it('onSetTime()', () => {
      const onSetTime = jest.fn();
      const wrapper = shallow(
        <CalendarDate highlight={new Date()} value={new Date()} onSetTime={onSetTime} time />
      );
      const newDate = new Date(2000, 11, 11, 0, 0, 0);
      wrapper
        .find('TimePicker')
        .props()
        .onSetTime(newDate);

      expect(onSetTime).toBeCalledWith(newDate);
    });

    it('onChangeMonth()', () => {
      const value = new Date(2000, 11, 11);
      const onChange = jest.fn();
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} onChange={onChange} show="mm" />
      );
      const newDate = setMonth(value, 3);
      wrapper
        .find('TableSelect')
        .props()
        .onChange(3);

      expect(wrapper.state('show')).toBe('calendar');
      expect(onChange).toBeCalledWith(newDate);
    });

    it('onChangeYear()', () => {
      const value = new Date(2019, 11, 11);
      const onChange = jest.fn();
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} onChange={onChange} show="yy" />
      );
      const newDate = setYear(value, 2018);
      wrapper
        .find('TableSelect')
        .props()
        .onChange(2018);

      expect(onChange).toBeCalledWith(newDate);
      expect(wrapper.state('show')).toBe('mm');
    });

    it('onChangeDecade()', () => {
      const value = new Date(2029, 11, 11);
      const onChange = jest.fn();
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} onChange={onChange} show="yy10" />
      );
      const newDate = setYear(value, 2048);
      wrapper
        .find('TableSelect')
        .props()
        .onChange(2048);

      expect(onChange).toBeCalledWith(newDate);
      expect(wrapper.state('show')).toBe('yy');
    });

    it('showCalendar()', () => {
      const wrapper = shallow(<CalendarDate highlight={new Date()} show="mm" value={new Date()} />);
      wrapper
        .find('Button')
        .props()
        .onClick();

      expect(wrapper.state('show')).toBe('calendar');
    });

    describe('showMonthTable()', () => {
      it('click month in CalendarMonthGrid', () => {
        const wrapper = shallow(<CalendarDate highlight={new Date()} value={new Date()} />);
        wrapper
          .find('CalendarMonthGrid')
          .props()
          .onClickMonth();

        expect(wrapper.state('show')).toBe('mm');
      });

      it('click back in years TableSelect', () => {
        const wrapper = shallow(
          <CalendarDate highlight={new Date()} value={new Date()} show="yy" />
        );
        wrapper
          .find('Button')
          .props()
          .onClick();

        expect(wrapper.state('show')).toBe('mm');
      });
    });

    describe('showYearTable()', () => {
      it('click year months TableSelect', () => {
        const wrapper = shallow(
          <CalendarDate highlight={new Date()} value={new Date()} show="mm" />
        );
        wrapper.find('span').simulate('click');

        expect(wrapper.state('show')).toBe('yy');
      });

      it('click back in 10YY TableSelect', () => {
        const wrapper = shallow(
          <CalendarDate highlight={new Date()} value={new Date()} show="yy10" />
        );
        wrapper
          .find('Button')
          .props()
          .onClick();

        expect(wrapper.state('show')).toBe('yy');
      });
    });

    it('showDecadeTable()', () => {
      const wrapper = shallow(<CalendarDate highlight={new Date()} value={new Date()} show="yy" />);
      wrapper.find('span').simulate('click');

      expect(wrapper.state('show')).toBe('yy10');
    });

    // TODO: TypeError: Cannot read property 'props' of undefined
    it.skip('incrementMonth()', () => {
      const onChange = jest.fn();
      const value = new Date(1994, 6, 0);
      const newDate = addMonths(value, 1);
      const wrapper = shallow(<CalendarDate value={value} onChange={onChange} />);
      wrapper
        .find('button')
        .at(1)
        .simulate('click');

      expect(onChange).toBeCalledWith(newDate);
    });

    it.skip('decrementMonth()', () => {
      const onChange = jest.fn();
      const value = new Date(1995, 8, 0);
      const newDate = addMonths(value, -1);
      const wrapper = shallow(<CalendarDate highlight={value} value={value} onChange={onChange} />);
      wrapper
        .find('button')
        .at(0)
        .simulate('click');

      expect(onChange).toBeCalledWith(newDate);
    });

    it('increment10Years()', () => {
      const onChange = jest.fn();
      const value = new Date(1915, 8, 0);
      const newDate = addYears(value, 10);
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} show="yy" onChange={onChange} />
      );
      wrapper
        .find('button')
        .at(1)
        .simulate('click');

      expect(onChange).toBeCalledWith(newDate);
    });

    it('decrement10Years()', () => {
      const onChange = jest.fn();
      const value = new Date(1965, 8, 0);
      const newDate = addYears(value, -10);
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} show="yy" onChange={onChange} />
      );
      wrapper
        .find('button')
        .at(0)
        .simulate('click');

      expect(onChange).toBeCalledWith(newDate);
    });

    it('incrementYears()', () => {
      const onChange = jest.fn();
      const value = new Date(1997, 8, 0);
      const newDate = addYears(value, 1);
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} show="mm" onChange={onChange} />
      );
      wrapper
        .find('button')
        .at(1)
        .simulate('click');

      expect(onChange).toBeCalledWith(newDate);
    });

    it('decrementYears()', () => {
      const onChange = jest.fn();
      const value = new Date(1977, 8, 0);
      const newDate = addYears(value, -1);
      const wrapper = shallow(
        <CalendarDate highlight={value} value={value} show="mm" onChange={onChange} />
      );
      wrapper
        .find('button')
        .at(0)
        .simulate('click');

      expect(onChange).toBeCalledWith(newDate);
    });
  });

  it('showDecade()', () => {
    const value = new Date(1477, 8, 0);
    const wrapper = shallow(<CalendarDate highlight={value} value={value} show="yy" />);
    wrapper.instance().showDecade();

    expect(wrapper.find('span').text()).toEqual('1470 - 1479');
  });

  it('getMonthOptions()', () => {
    const value = new Date(1887, 8, 0);
    const wrapper = shallow(<CalendarDate highlight={value} value={value} />);
    wrapper.instance().getMonthOptions();

    expect(wrapper.state('monthsOptions')).toMatchSnapshot();
  });

  it('getYearOptions()', () => {
    const value = new Date(1767, 8, 0);
    const wrapper = shallow(<CalendarDate highlight={value} value={value} show="yy" />);
    wrapper.instance().getYearOptions(value);
    const { options } = wrapper.find('TableSelect').props();

    expect(options).toMatchSnapshot();
  });

  it('getYearDecadeOptions()', () => {
    const value = new Date(1767, 8, 0);
    const wrapper = shallow(<CalendarDate highlight={value} value={value} show="yy" />);
    const options = wrapper.instance().getYearDecadeOptions(value);

    expect(options).toMatchSnapshot();
  });

  describe('getBtnStyle()', () => {
    const value = new Date(1767, 8, 0);
    it('if leftArrow && !rightArrow', () => {
      const wrapper = shallow(<CalendarDate value={value} leftArrow />);
      const btnStyle = wrapper.instance().getBtnStyle();

      expect(btnStyle).toEqual({ justifyContent: 'flex-start' });
    });

    it('if !leftArrow && rightArrow', () => {
      const wrapper = shallow(<CalendarDate value={value} rightArrow />);
      const btnStyle = wrapper.instance().getBtnStyle();

      expect(btnStyle).toEqual({ justifyContent: 'flex-end' });
    });

    it('else', () => {
      const wrapper = shallow(<CalendarDate value={value} />);
      const btnStyle = wrapper.instance().getBtnStyle();

      expect(btnStyle).toEqual({ justifyContent: 'space-between' });
    });
  });
});
