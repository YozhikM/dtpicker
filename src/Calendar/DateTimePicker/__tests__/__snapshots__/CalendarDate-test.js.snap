// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarDate componentWillReceiveProps() 1`] = `1994-07-31T17:00:00.000Z`;

exports[`CalendarDate getMonthOptions() 1`] = `
Array [
  Object {
    "name": "Jan",
    "value": 0,
  },
  Object {
    "name": "Feb",
    "value": 1,
  },
  Object {
    "name": "Mar",
    "value": 2,
  },
  Object {
    "name": "Apr",
    "value": 3,
  },
  Object {
    "name": "May",
    "value": 4,
  },
  Object {
    "name": "Jun",
    "value": 5,
  },
  Object {
    "name": "Jul",
    "value": 6,
  },
  Object {
    "name": "Aug",
    "value": 7,
  },
  Object {
    "name": "Sep",
    "value": 8,
  },
  Object {
    "name": "Oct",
    "value": 9,
  },
  Object {
    "name": "Nov",
    "value": 10,
  },
  Object {
    "name": "Dec",
    "value": 11,
  },
]
`;

exports[`CalendarDate getYearDecadeOptions() 1`] = `
Array [
  Object {
    "name": "1730 1739",
    "value": 1730,
  },
  Object {
    "name": "1740 1749",
    "value": 1740,
  },
  Object {
    "name": "1750 1759",
    "value": 1750,
  },
  Object {
    "name": "1760 1769",
    "value": 1760,
  },
  Object {
    "name": "1770 1779",
    "value": 1770,
  },
]
`;

exports[`CalendarDate getYearOptions() 1`] = `
Array [
  Object {
    "name": "1760",
    "value": 1760,
  },
  Object {
    "name": "1761",
    "value": 1761,
  },
  Object {
    "name": "1762",
    "value": 1762,
  },
  Object {
    "name": "1763",
    "value": 1763,
  },
  Object {
    "name": "1764",
    "value": 1764,
  },
  Object {
    "name": "1765",
    "value": 1765,
  },
  Object {
    "name": "1766",
    "value": 1766,
  },
  Object {
    "name": "1767",
    "value": 1767,
  },
  Object {
    "name": "1768",
    "value": 1768,
  },
  Object {
    "name": "1769",
    "value": 1769,
  },
]
`;

exports[`CalendarDate render basic 1`] = `
<div>
  <div
    style={
      Object {
        "justifyContent": "space-between",
      }
    }
  />
  <CalendarMonthGrid
    highlight={2010-11-09T18:00:00.000Z}
    onClickMonth={[Function]}
    onSetDate={[Function]}
    value={2010-11-09T18:00:00.000Z}
  />
</div>
`;

exports[`CalendarDate render if decade table 1`] = `
<div>
  <div>
    <span>
        
      <br />
        
    </span>
    <TableSelect
      cols={4}
      onChange={[Function]}
      options={
        Array [
          Object {
            "name": "2000 2009",
            "value": 2000,
          },
          Object {
            "name": "2010 2019",
            "value": 2010,
          },
          Object {
            "name": "2020 2029",
            "value": 2020,
          },
          Object {
            "name": "2030 2039",
            "value": 2030,
          },
          Object {
            "name": "2040 2049",
            "value": 2040,
          },
          Object {
            "name": "2050 2059",
            "value": 2050,
          },
          Object {
            "name": "2060 2069",
            "value": 2060,
          },
          Object {
            "name": "2070 2079",
            "value": 2070,
          },
        ]
      }
      value={2040}
    />
    <Button
      onClick={[Function]}
      outline={true}
      type="button"
      xs={true}
    >
      Назад
    </Button>
  </div>
</div>
`;

exports[`CalendarDate render if disabled leftArrow 1`] = `
<div>
  <div
    style={
      Object {
        "justifyContent": "space-between",
      }
    }
  />
  <CalendarMonthGrid
    highlight={2007-07-04T18:00:00.000Z}
    onClickMonth={[Function]}
    onSetDate={[Function]}
    value={2007-07-04T18:00:00.000Z}
  />
</div>
`;

exports[`CalendarDate render if disabled rightArrow 1`] = `
<div>
  <div
    style={
      Object {
        "justifyContent": "space-between",
      }
    }
  />
  <CalendarMonthGrid
    highlight={1984-05-03T17:00:00.000Z}
    onClickMonth={[Function]}
    onSetDate={[Function]}
    value={1984-05-03T17:00:00.000Z}
  />
</div>
`;

exports[`CalendarDate render if month table 1`] = `
<div>
  <div>
    <div>
      <button
        onClick={[Function]}
      >
        <SvgIcon
          file="arrow-left"
        />
      </button>
      <span
        onClick={[Function]}
      >
        2014
      </span>
      <button
        onClick={[Function]}
      >
        <SvgIcon
          file="arrow-right"
        />
      </button>
    </div>
    <TableSelect
      cols={4}
      onChange={[Function]}
      options={
        Array [
          Object {
            "name": "Jan",
            "value": 0,
          },
          Object {
            "name": "Feb",
            "value": 1,
          },
          Object {
            "name": "Mar",
            "value": 2,
          },
          Object {
            "name": "Apr",
            "value": 3,
          },
          Object {
            "name": "May",
            "value": 4,
          },
          Object {
            "name": "Jun",
            "value": 5,
          },
          Object {
            "name": "Jul",
            "value": 6,
          },
          Object {
            "name": "Aug",
            "value": 7,
          },
          Object {
            "name": "Sep",
            "value": 8,
          },
          Object {
            "name": "Oct",
            "value": 9,
          },
          Object {
            "name": "Nov",
            "value": 10,
          },
          Object {
            "name": "Dec",
            "value": 11,
          },
        ]
      }
      value={11}
    />
    <Button
      onClick={[Function]}
      outline={true}
      type="button"
      xs={true}
    >
      Назад
    </Button>
  </div>
</div>
`;

exports[`CalendarDate render if time 1`] = `
<div>
  <div
    style={
      Object {
        "justifyContent": "space-between",
      }
    }
  />
  <CalendarMonthGrid
    highlight={2005-07-06T18:00:00.000Z}
    onClickMonth={[Function]}
    onSetDate={[Function]}
    value={2005-07-06T18:00:00.000Z}
  />
  <TimePicker
    onSetTime={[Function]}
  />
</div>
`;

exports[`CalendarDate render if year table 1`] = `
<div>
  <div>
    <div>
      <button
        onClick={[Function]}
      >
        <SvgIcon
          file="arrow-left"
        />
      </button>
      <span
        onClick={[Function]}
      >
        2020 - 2029
      </span>
      <button
        onClick={[Function]}
      >
        <SvgIcon
          file="arrow-right"
        />
      </button>
    </div>
    <TableSelect
      cols={4}
      onChange={[Function]}
      options={
        Array [
          Object {
            "name": "2020",
            "value": 2020,
          },
          Object {
            "name": "2021",
            "value": 2021,
          },
          Object {
            "name": "2022",
            "value": 2022,
          },
          Object {
            "name": "2023",
            "value": 2023,
          },
          Object {
            "name": "2024",
            "value": 2024,
          },
          Object {
            "name": "2025",
            "value": 2025,
          },
          Object {
            "name": "2026",
            "value": 2026,
          },
          Object {
            "name": "2027",
            "value": 2027,
          },
          Object {
            "name": "2028",
            "value": 2028,
          },
          Object {
            "name": "2029",
            "value": 2029,
          },
        ]
      }
      value={2022}
    />
    <Button
      onClick={[Function]}
      outline={true}
      type="button"
      xs={true}
    >
      Назад
    </Button>
  </div>
</div>
`;
